src/foam/core/Action.js

```diff
@@ -113,80 +112,99 @@ foam.CLASS({
       value: null
     },
     {
-      class: 'Boolean',
-      name: 'permissionRequired',
-      documentation: 'When set to true, a permission is needed to execute this action.'
-    },
-    {
-      class: 'FObjectProperty',
-      of: 'foam.nanos.auth.ActionPermissionConfiguration',
-      name: 'permissionConfig',
-      factory: function() {
-        if ( this.permissionRequired ) {
-          var defaultPermission = this.sourceCls_.id + '.permission.' + this.name;
-          return this.ActionPermissionConfiguration.create({
-            available: [defaultPermission],
-            enabled: [defaultPermission]
-          });
-        }
-      }
+      class: 'StringArray',
+      name: 'availablePermissions',
+      documentation: `Permissions required for the action to be available.
+If empty than no permissions are required.`
+    },
+    {
+      class: 'StringArray',
+      name: 'enabledPermissions',
+      documentation: `Permissions required for the action to be enabled.
+If empty than no permissions are required.`,
     }
   ],
 
   methods: [
-    function andSlots(a, b) {
-      return this.ExpressionSlot.create({
-        args: [ a, b ],
+    function addPermissionsCheck_(x, slot, data, permissions) {
+      // Decorates an isEnabled/isAvailable slot with a permission
+      // check if appropriate.
+
+      // If no auth service, or no permissions to check then nothing to do.
+      if ( ! x.auth || ! permissions.length )
+        return slot;
+
+      return foam.core.ExpressionSlot.create({
+        args: [
+          foam.core.PromiseSlot.create({
+            promise: Promise.all(permissions.map(p => x.auth.check(null, p))).
+              then(function(perms) {
+                return perms.every(p => p);
+              })
+          }),
+          slot
+        ],
         code: function(a, b) {
           return a && b;
         }
       });
     },
 
-    function andSlotAndPromise(slot, promise) {
-      return this.andSlots(slot, this.PromiseSlot.create({
-        promise: promise
-      }));
-    },
+    function createSlotFor_(x, data, expression, permissions) {
+      // Handle old code that might try to pass data as a slot
+      if ( foam.core.Slot.isInstance(data) ) {
+        console.warn("Action createIsEnabled$ and createIsAvailable$ does not support data as a slot.");
+        data = data.get();
+      }
 
-    function checkPermission(x, permission) {
-      if ( ! this.permissionConfig || ! x.auth ) return Promise.resolve(true);
-      return x.auth.check(null, permission);
-    },
+      // Helper method for creating isEnabled/isAvailable slots.
 
-    function isEnabledFor(data) {
-      return this.isEnabled
-        ? foam.Function.withArgs(this.isEnabled, data)
-        : true;
-    },
+      var slot = expression ?
+          foam.core.ExpressionSlot.create({
```

Comment:
I think you can just do data.slot(expression).

Reply:
Done

Reply:
It still shows the old code.